Allow user input for satellite download + utility/helper functions

vikasnataraja opened the pull request:
This PR has two changes:

1. Additions to `util/util.py` include the `re_extend` and `get_lon2lat2` functions which allows the creation of same-sized images by re-extending the given extent. However, this requires the use of `geopy` and `math` as new package imports which might complicate things, so feel free to discard this change.

2. Changes to `examples/02_modis_rad-sim.py` file to: 
    - allow users to pass both satellite names (Aqua and Terra) as inputs to the `satellite_download` class instead of always downloading Aqua data. This is handled by pre-setting the dataset prefixes before calling `download_laads_https`. 
    - include a few more stdout user notifications

vikasnataraja committed:
add new functions for same image size extent gridding

vikasnataraja committed:
add support for both terra and aqua as user input+ stdout statements

hong-chen commented on er3t/util/util.py:
```diff
@@ -1,13 +1,15 @@
 import requests
 import urllib.request
 from io import StringIO
+import geopy.distance as gdist
```
If `geopy` is only used to calculate distance between two locations, `cartopy` should contain some similar functionalities (I will look into this). We should avoid introducing new dependent libraries (unless we really have to) to keep `er3t` compact and lightweight.

hong-chen commented on er3t/util/util.py:
```diff
@@ -404,6 +406,72 @@ def get_doy_tag(date, day_interval=8):
     return doy_tag
 
 
+def get_lon2lat2(lon1, lat1, bearing, distance, use_great_circle=True):
```
How about rename this function to `cal_lonlat_by_dist` so it is more self-explanatory? Maybe array operation is possible (e.g., input array of `lon1`, `lat1` and output array `lon2`, `lat2`) by replacing `math` functions with `numpy` functions.

vikasnataraja commented:
Agreed, let me work on this tomorrow so we don’t introduce additional libraries especially just for simple functions. We can probably replace trigonometric functions with those in numpy but I will test it out first and push a commit to this PR. And rename it accordingly.

hong-chen commented:
> Agreed, let me work on this tomorrow so we don’t introduce additional libraries especially just for simple functions. We can probably replace trigonometric functions with those in numpy but I will test it out first and push a commit to this PR. And rename it accordingly.

Don't spend too much time on it - it is already good enough. I was just thinking out loud. Very nice comments/descriptions for the functions!

hong-chen reviewed (approved):
The changes are nicely done.
Eventually (being futuristic at this point), command line tools (executables) will be created under `bin` directory. For satellite data download, my vision is to have an executable (something like `download_modis`) so users can download data by typing in a command line something like
```
download_modis --date=20230101 --lon=0.0 --lat=0.0 --sat=aqua --tag=mod03 --fdir='tmp-data'
```
With current functions included in `er3t`, this can be done easily. If you are interested, you can initiate such implementation (I will help you with it).

vikasnataraja commented:
That is actually how I have it running on my end. I have a command line setup with `argparse` for both downloading and creating my training data. I find this very helpful because parameters like resolution, width, height can be easily changed. I will refine that a little bit and push a separate PR. 

For this PR, I will push the modifications we talked about yesterday and merge with `dev`.

vikasnataraja committed:
add haversine fn, remove geopy, and rename fns

vikasnataraja commented:
I was able to replace `math` with functions in `numpy`. The geodetic distance that `geopy` uses is harder to replace, however. I couldn't find a direct replacement in `cartopy`. I figure one should not be using geodetic distance with MODIS anyway but we should revisit this issue at a later point if that is deemed useful for other functionalities. So, I removed `geopy` import and its functionalities.  I have now also added a function to calculate Haversine distance which should be very close to geodetic distance. I am now setting the radius of the Earth as a constant (which is what `geopy` also does in their code). Feel free to move things around for placement.

Latest commit: https://github.com/hong-chen/er3t/commit/8d024e8c22d59d0eb6541af84a4f9ec3eea9a666

vikasnataraja committed:
Merge branch 'dev' into dev_vikas
